src/server/api/contactsApi.js

```diff
@@ -0,0 +1,165 @@
+import { r } from '../models'
+import _ from 'lodash'
+import { getTimezoneByZip, getOptOutSubQuery } from '../../workers/jobs'
+
+import { getValidatedData } from '../../lib'
+
+function campaignaStatusShortCircuit(campaign, res) {
+  let message = ''
+  if (campaign.is_archivedj) {
+    message = 'Campaign is archived'
+  } else if (campaign.is_started) {
+    message = 'Campaign is started'
+  }
+
+  if (message) {
+    res.writeHead(403)
+    res.end(message)
+    return true
+  }
+
+  return false
+}
+
+// TODO(lperson) enforce that this is not reentrant
+// TODO(lperson) enforce auth
+export default async function contactsApi(req, res) {
+  const orgId = req.params.orgId
+  const campaignId = req.params.campaignId
+
+  if (!['GET', 'DELETE', 'POST'].includes(req.method)) {
+    res.writeHead(405, { Allow: 'GET, POST, DELETE' })
+    res.end('Not allowed')
+    return
+  }
+
+  const campaignArray = await r
+    .knex('campaign')
+    .select()
+    .where({ organization_id: orgId, id: campaignId })
+
+  if (campaignArray.length !== 1) {
+    res.writeHead(404)
+    res.end('Not found')
+    return
+  }
+
+  const campaign = campaignArray[0]
+
+  let resp = {}
+
+  switch (req.method) {
+    case 'GET':
+      const count = await r
+        .knex('campaign_contact')
+        .where({ campaign_id: campaignId })
+        .count('*')
+
+      resp = { contacts: { count: parseInt(count[0].count) } }
+      res.writeHead(200, { 'Content-Type': 'application/json' })
+      res.end(JSON.stringify(resp))
+      break
+
+    case 'DELETE':
+      if (campaignaStatusShortCircuit(campaign, res)) {
+        break
+      }
+
+      const deleted = await r
+        .knex('campaign_contact')
+        .delete()
+        .where({ campaign_id: campaignId })
+
+      resp = { contacts: { deleted } }
+      res.writeHead(200, { 'Content-Type': 'application/json' })
+      res.end(JSON.stringify(resp))
+      break
+
+    case 'POST':
+      if (campaignaStatusShortCircuit(campaign, res)) {
+        break
+      }
+
+      const { validationStats, validatedData } = getValidatedData(req.body, [])
+
+      const successResponse = {
+        invalid: _.concat(
+          validationStats.invalidCellRows,
+          validationStats.missingCellRows
+        ),
+        dupesInBatch: validationStats.dupeCount,
+        numberSubmitted: req.body.length
+      }
+
+      const standardFields = [
+        'first_name',
+        'last_name',
+        'cell',
+        'external_id',
+        'zip'
+      ]
+
+      const contactsToSavePromises = validatedData.map(async contact => {
+        const contactToSave = _.pick(contact, standardFields)
+        const customFields = _.omit(contact, standardFields)
+        contactToSave.custom_fields = JSON.stringify(customFields)
+        contactToSave.timezone_offset = await getTimezoneByZip(
```

Comment:
we are doing this in the db query, too.  Please test for a timezone_offset column and use it instead of doing the zip lookup if it exists.

Reply:
👍